pennylane/devices/qubit/simulate.py

```diff
@@ -46,6 +47,59 @@
 }
 
 
+def _valid_flex_int(s):
+    """Returns True if s is a non-negative integer."""
+    return isinstance(s, int) and s >= 0
+
+
+def _valid_flex_tuple(s):
+    """Returns True if s is a tuple of the form (shots, copies)."""
+    return (
+        isinstance(s, tuple)
+        and len(s) == 2
+        and _valid_flex_int(s[0])
+        and isinstance(s[1], int)
+        and s[1] > 0
+    )
+
+
+class _FlexShots(qml.measurements.Shots):
```

Comment:
Do we need a new class or could we just modify `qml.measurements.Shots`?

Reply:
`Shots` is a frozen class, so we cannot mutate it in-place. The only way to create `Shots` with zero shots is to pass such a shot vector to the constructor, which won't work as all the shots related validation is handled by this class. This could open up the possibility of users passing in 0 shots without errors, leading to bad behaviour down the line.